README.md
# splash_screen_view

A flutter package which contains a collection of Splash Screen example for your app to display logo and different text style.

# Installing

### 1. Depend on it
Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  splash_screen_view: ^1.0.1
```

### 2. Install it

You can install packages from the command line:

with `pub`:

```css
$ pub get
```

with `Flutter`:

```css
$ flutter packages get
```

### 3. Import it

Now in your `Dart` code, you can use:

```dart
import 'package:splash_screen_view/SplashScreenView.dart';
```


# Usage


## Simple Logo

<img src="https://github.com/SandipVKalola/splash_screen/blob/master/display/OnlyImage.gif?raw=true" align = "right" height = "400px">

```dart
SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 200,
      imageSrc: "logo.png",
      backgroundColor: Colors.white,
    );
```
[Where]:
 - home (required)- Name of target screen which you want to display after completion of splash screen milliseconds.
 - duration (required) - Delay between the splash screen and target screen. Provider Duration in millisecond.
 - imageSrc (required)- Assets path for your logo which your want to display on splash screen.
 - imageSize - Size of your logo. By default it is 150.
 - backgroundColor - Background color of splash screen. By default it is white color.


## Logo with Simple Text

<img src="https://github.com/SandipVKalola/splash_screen/blob/master/display/Normal.gif?raw=true" align = "right"  height = "400px">

```dart
SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Normal Text",
      textType: TextType.NormalText,
      textStyle: TextStyle(
        fontSize: 30.0,
      ),
      backgroundColor: Colors.white,
    );
```
[Where]:
 - text - Text which you want to show below logo.
 - textType - Gives text type as TextType.NormalText
 - textStyle - Gives TextStyle to the text strings.
**Note:** Give `imageSrc` as blank If you want only text on your splash screen.

## Colorize Animated Text

<img src="https://github.com/SandipVKalola/splash_screen/blob/master/display/ColorizeAnimatedText.gif?raw=true" align = "right"  height = "400px">

```dart
SplashScreenView(
      home: SecondScreen(),
      duration: 5000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Colorize Text",
      textType: TextType.ColorizeAnimationText,
      textStyle: TextStyle(
        fontSize: 40.0,
      ),
      colors: [
        Colors.purple,
        Colors.blue,
        Colors.yellow,
        Colors.red,
      ],
      backgroundColor: Colors.white,
    );
```
[Where]:
 - textType - Gives text type as TextType.ColorizeAnimationText
 - colors - Set the colors for the gradient animation of the text. Give List with values of [Color] in it.
**Note:** `colors` list should contains at least two values.


## Scale Animated Text

<img src="https://github.com/SandipVKalola/splash_screen/blob/master/display/ScaleAnimatedText.gif?raw=true" align = "right"  height = "400px">

```dart
SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Scale Animated Text",
      textType: TextType.ScaleAnimatedText,
      textStyle: TextStyle(
        fontSize: 30.0,
      ),
      backgroundColor: Colors.white,
    );
```
[Where]:
 - textType - Gives text type as TextType.ScaleAnimatedText


## Typer Animated Text

<img src="https://github.com/SandipVKalola/splash_screen/blob/master/display/TyperAnimatedText.gif?raw=true" align = "right"  height = "400px">

```dart
SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Typer Animated Text",
      textType: TextType.TyperAnimatedText,
      textStyle: TextStyle(
        fontSize: 30.0,
      ),
      backgroundColor: Colors.white,
    );
```
[Where]:
 - textType - Gives text type as TextType.TyperAnimatedText

# Bugs or Requests

If you encounter any problems feel free to open an [issue](https://github.com/SandipVKalola/splash_screen/issues/new?template=bug_report.md). If you feel the library is missing a feature, please raise a [ticket](https://github.com/SandipVKalola/splash_screen/issues/new?template=feature_request.md) on GitHub and I'll look into it. Pull request are also welcome.


## Github example link
https://github.com/SandipVKalola/splash_screen

# Donate
> If you found this project helpful or you learned something from the source code and want to thank me, consider buying me a cup of :coffee:
>
> - [PayPal](https://paypal.me/sandipkalola)

# License
Splash Screen is licensed under `MIT license`. View [license](https://github.com/SandipVKalola/splash_screen/blob/master/LICENSE).


CHANGELOG.md

## [1.1.0] - 30 Jan 2023
* Migrated to nullsfatey

## [1.0.3] - 28 Sep 2020

* Update README.md File

## [1.0.2] - 28 Sep 2020

* Update README.md File

## [1.0.1] - 27 Sep 2020

* Added: Colorize Animated Text Effect
* Added: Scale Animated Text Effect
* Added: Typer Animated Text Effect

## [1.0.0] - 27 Sep 2020

* initial release.


lib/SplashScreenView.dart
library splash_screen_view;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:splash_screen_view/ColorizeAnimatedText.dart';
import 'package:splash_screen_view/ScaleAnimatedText.dart';
import 'package:splash_screen_view/TyperAnimatedText.dart';

enum TextType {
  ColorizeAnimationText,
  TyperAnimatedText,
  ScaleAnimatedText,
  NormalText,
}

// ignore: must_be_immutable
class SplashScreenView extends StatefulWidget {
  /// The [Widget] Name of target screen which you want to display after completion of splash screen milliseconds.
  late Widget _home;

  ///  [String] Asset path of your logo image.
  late String _imageSrc = "";

  ///  [String] that would be displayed on below logo.
  late String _text;

  /// Select [TextType] of your text.
  ///
  /// By default it is NormalText.
  late TextType _textType;

  /// Gives [TextStyle] to the text strings.
  late TextStyle _textStyle;

  /// The [Duration] of the delay between the apparition of each characters
  ///
  /// By default it is set to 3000 milliseconds.
  late int _duration = 3000;

  ///  [int] Size of your image logo.
  ///
  /// By default it is set to 150.
  late int _logoSize = 150;

  ///  [Color] Background Color of your splash screen.
  /// By default it is set to white.
  Color _backgroundColor = Colors.white;

  /// Set the colors for the gradient animation of the text.
  ///
  /// The [List] should contain at least two values of [Color] in it.
  /// By default it is set to red and black.
  late List<Color> _colors;

  Future<dynamic>? _future;

  late bool _waitUntilPop;

  late bool _redirectUponPop;

  String? _loadingText;

  String? _errorText;

  String? _successText;

  Widget? _homeOnError;

  double _defaultTextFontSize = 20;

  SplashScreenView(
      {required String imageSrc,
      required Widget home,
      required int duration,
      int? imageSize,
      textStyle,
      colors,
      textType,
      backgroundColor,
      text,
      Future<dynamic>? future,
      bool waitUntilPop = false,
      bool redirectUponPop = true,
      String? successText,
      String? loadingText,
      String? errorText,
      Widget? homeOnError}) {
    assert(duration != null);
    assert(home != null);
    assert(imageSrc != null);

    _imageSrc = imageSrc;
    _home = home;
    _duration = duration;
    _colors = colors;
    _text = text;
    _textStyle = textStyle;
    _logoSize = imageSize ?? 10;
    _backgroundColor = backgroundColor;
    _textType = textType;

    _future = future;
    _waitUntilPop = waitUntilPop;
    _redirectUponPop = redirectUponPop;
    _loadingText = loadingText ?? "Loading...";
    _errorText = errorText ?? "Error!";
    _homeOnError = homeOnError ?? Container(child: Text("HOME ERROR"));
    _successText = successText ?? "SUCCESS";
  }

  @override
  _SplashScreenViewState createState() => _SplashScreenViewState();
}

class _SplashScreenViewState extends State<SplashScreenView>
    with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _animation;

  @override
  void initState() {
    super.initState();
    if (widget._duration < 1000) widget._duration = 3000;

    if (widget._textType == TextType.TyperAnimatedText) {
      _animationController = new AnimationController(
          vsync: this, duration: Duration(milliseconds: 100));
      _animation = Tween(begin: 0.0, end: 1.0).animate(CurvedAnimation(
          parent: _animationController, curve: Curves.easeInCirc));
      _animationController.forward();
    } else {
      _animationController = new AnimationController(
          vsync: this, duration: Duration(milliseconds: 1000));
      _animation = Tween(begin: 0.0, end: 1.0).animate(CurvedAnimation(
          parent: _animationController, curve: Curves.easeInCirc));
      _animationController.forward();
    }

    if(widget._future == null){
      Future.delayed(Duration(milliseconds: widget._duration)).then((value) => 
        Navigator.of(context).pushReplacement(
          CupertinoPageRoute(builder: (BuildContext context) => widget._home))
      );
    }

  }

  @override
  void dispose() {
    super.dispose();
    _animationController.reset();
  }

  @override
  Widget build(BuildContext context) {

    return WillPopScope(
      onWillPop: () async {
        if(widget._waitUntilPop){
          if(widget._redirectUponPop){
            Navigator.of(context).pushReplacement(
              CupertinoPageRoute(builder: (BuildContext context) => widget._home));
          }
        }
        return false;
      },
      child: Scaffold(
        backgroundColor: widget._backgroundColor,
        body: Container(
          width: MediaQuery.of(context).size.width,
          height: MediaQuery.of(context).size.height,
          child: FadeTransition(
            opacity: _animation,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: <Widget>[
                (widget._imageSrc != null)
                    ? Image.asset(
                        widget._imageSrc,
                        height: (widget._logoSize != null)
                            ? widget._logoSize.toDouble()
                            : 150,
                      )
                    : SizedBox(
                        height: 1,
                      ),
                (widget._future != null)
                ?
                FutureBuilder(
                  future: widget._future,
                  builder: (BuildContext context, AsyncSnapshot snapshot) {

                    String text;

                    if(snapshot.hasData){
                      text = widget._text + "\n\n" + widget._successText!;
                      Future.delayed(Duration(milliseconds: 3000)).then((value) => 
                        Navigator.of(context).pushReplacement(
                          CupertinoPageRoute(builder: (BuildContext context) => widget._home))
                      );
                      return Padding(
                        padding: const EdgeInsets.only(right: 10, left: 10, top: 20),
                        child: getTextWidget(text),
                      );
                      // Future.delayed(Duration(milliseconds: widget._duration)).then((value) {
                      //   Navigator.of(context).pushReplacement(
                      //       CupertinoPageRoute(builder: (BuildContext context) => widget._home));
                      // });
                    } else if(snapshot.hasError){
                      text = widget._text + "\n\n" + widget._errorText!;
                      Future.delayed(Duration(milliseconds: 3000)).then((value) => 
                        Navigator.of(context).pushReplacement(
                          CupertinoPageRoute(builder: (BuildContext context) => widget._homeOnError!))
                      );
                      return Padding(
                        padding: const EdgeInsets.only(right: 10, left: 10, top: 20),
                        child: getTextWidget(text),
                      );
                    } else {
                      text = widget._text + "\n\n" + widget._loadingText!;
                      return Padding(
                        padding: const EdgeInsets.only(right: 10, left: 10, top: 20),
                        child: getTextWidget(text),
                      );
                    }

                  },
                )
                :
                Padding(
                  padding: const EdgeInsets.only(right: 10, left: 10, top: 20),
                  child: getTextWidget(widget._text),
                )
              ],
            ),
          ),
        ),
      ),
    );
  }

  Widget getTextWidget(String text) {
    if (widget._text != null) {
      print("Not Blank");
      switch (widget._textType) {
        case TextType.ColorizeAnimationText:
          return ColorizeAnimatedText(
            text: text,
            speed: Duration(milliseconds: 1000),
            textStyle: (widget._textStyle != null)
                ? widget._textStyle
                : TextStyle(fontSize: widget._defaultTextFontSize),
            colors: (widget._colors != null)
                ? widget._colors
                : [
                    Colors.red,
                    Colors.black,
                    Colors.red,
                    Colors.black,
                  ],
          );
        case TextType.NormalText:
          return Text(
            text,
            style: (widget._textStyle != null)
                ? widget._textStyle
                : TextStyle(fontSize: widget._defaultTextFontSize),
          );
        case TextType.TyperAnimatedText:
          return TyperAnimatedText(
            text: text,
            speed: Duration(milliseconds: 100),
            textStyle: (widget._textStyle != null)
                ? widget._textStyle
                : TextStyle(fontSize: widget._defaultTextFontSize),
          );
        case TextType.ScaleAnimatedText:
          return ScaleAnimatedText(
            text: text,
            textStyle: (widget._textStyle != null)
                ? widget._textStyle
                : TextStyle(fontSize: widget._defaultTextFontSize),
          );
        default:
          return Text(
            text,
            style: (widget._textStyle != null)
                ? widget._textStyle
                : TextStyle(fontSize: widget._defaultTextFontSize),
          );
      }
    } else {
      print("Blank");
      return SizedBox(
        width: 1,
      );
    }
  }
}


lib/ScaleAnimatedText.dart
import 'package:flutter/material.dart';

class ScaleAnimatedText extends StatefulWidget {
  /// [String] that would be displayed subsequently in the animation.
  final String text;

  /// Gives [TextStyle] to the text strings.
  final TextStyle? textStyle;

  /// Override the [Duration] of the animation by setting the duration parameter.
  ///
  /// This will set the total duration for the animated widget.
  /// For example, if text = ["a", "b", "c"] and if you want that each animation
  /// should take 3 seconds then you have to set [duration] to 3 seconds.
  final Duration? duration;

  /// Set the scaling factor of the text for the animation.
  ///
  /// By default it is set to [double] value 0.5
  final double scalingFactor;

  const ScaleAnimatedText({
    Key? key,
    required this.text,
    this.textStyle,
    this.scalingFactor = 0.2,
    this.duration,
  }) : super(key: key);

  @override
  _ScaleTextState createState() => _ScaleTextState();
}

class _ScaleTextState extends State<ScaleAnimatedText>
    with TickerProviderStateMixin {
  late AnimationController _controller;
  late Animation<double> _fadeIn, _fadeOut, _scaleIn, _scaleOut;
  late Duration _duration;

  @override
  void initState() {
    super.initState();
    _duration = widget.duration ?? const Duration(milliseconds: 2300);
    setState(() {});
    _controller = AnimationController(
      duration: _duration,
      vsync: this,
    );

    _fadeIn = Tween<double>(begin: 0.0, end: 1.0).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(0.0, 0.8, curve: Curves.easeOut),
      ),
    );

    _fadeOut = Tween<double>(begin: 1.0, end: 0.0).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(0.8, 1.0, curve: Curves.easeOut),
      ),
    );

    _scaleIn = Tween<double>(begin: 0.0, end: 1.0).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(
          0.0,
          1.0,
          curve: Curves.easeOut,
        ),
      ),
    );
    _scaleOut = Tween<double>(begin: 1.0, end: 0.0).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(
          1.0,
          1.0,
          curve: Curves.easeIn,
        ),
      ),
    );
    _controller.forward();
  }

  @override
  void dispose() {
    _controller?.stop();
    _controller?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _controller,
      child: Text(
        widget.text,
        textAlign: TextAlign.center,
        style: widget.textStyle,
      ),
      builder: (BuildContext context, Widget? child) {
        return ScaleTransition(
          scale: _scaleIn.value != 1.0 ? _scaleIn : _scaleOut,
          child: Opacity(
            opacity: _fadeIn.value != 1.0 ? _fadeIn.value : _fadeOut.value,
            child: child,
          ),
        );
      },
    );
  }
}


lib/TyperAnimatedText.dart
import 'package:flutter/material.dart';

class TyperAnimatedText extends StatefulWidget {
  /// [String] that would be displayed subsequently in the animation.
  final String text;

  /// Gives [TextStyle] to the text strings.
  final TextStyle? textStyle;

  /// The [Duration] of the delay between the apparition of each characters
  ///
  /// By default it is set to 40 milliseconds.
  final Duration? speed;

  const TyperAnimatedText({
    Key? key,
    required this.text,
    this.textStyle,
    this.speed,
  }) : super(key: key);

  @override
  _TyperState createState() => _TyperState();
}

class _TyperState extends State<TyperAnimatedText>
    with TickerProviderStateMixin {
  late AnimationController _controller;
  late Animation _typingText;

  @override
  void initState() {
    super.initState();
    if (mounted) setState(() {});

    _controller = AnimationController(
      duration: (widget.speed ?? Duration(seconds: 3)) * widget.text.length,
      vsync: this,
    );

    _typingText =
        StepTween(begin: 0, end: widget.text.length).animate(_controller);

    _controller.forward();
  }

  @override
  void dispose() {
    _controller.stop();
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _controller,
      builder: (BuildContext context, Widget? child) {
        final int offset = widget.text.length < _typingText.value
            ? widget.text.length
            : _typingText.value;

        return Text(
          widget.text.substring(0, offset),
          textAlign: TextAlign.center,
          style: widget.textStyle,
        );
      },
    );
  }
}


lib/ColorizeAnimatedText.dart
import 'package:flutter/material.dart';

class ColorizeAnimatedText extends StatefulWidget {
  ///  [String] that would be displayed subsequently in the animation.
  final String text;

  /// Gives [TextStyle] to the text strings.
  final TextStyle? textStyle;

  /// The [Duration] of the delay between the apparition of each characters
  ///
  /// By default it is set to 500 milliseconds.
  final Duration? speed;

  /// Set the colors for the gradient animation of the text.
  ///
  /// The [List] should contain at least two values of [Color] in it.
  final List<Color> colors;

  const ColorizeAnimatedText({
    Key? key,
    required this.text,
    this.textStyle,
    required this.colors,
    this.speed,
  }) : super(key: key);

  @override
  _ColorizeTextState createState() => _ColorizeTextState();
}

class _ColorizeTextState extends State<ColorizeAnimatedText>
    with TickerProviderStateMixin {
  late AnimationController _controller;
  late Animation _colorShifter, _fadeIn, _fadeOut;
  late double _tuning;

  @override
  void initState() {
    super.initState();
    if (mounted) setState(() {});
    _controller = AnimationController(
      duration: (widget.speed ?? Duration(seconds: 3))* widget.text.length,
      vsync: this,
    );

    _tuning = (300.0 * widget.colors.length) *
        ((widget.textStyle?.fontSize ?? 12) / 24.0) *
        0.75 *
        (widget.text.length / 15.0);

    _fadeIn = Tween<double>(begin: 0.0, end: 1.0).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(0.0, 0.1, curve: Curves.easeOut),
      ),
    );

    _fadeOut = Tween<double>(begin: 1.0, end: 1.0).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(0.9, 1.0, curve: Curves.easeIn),
      ),
    );

    _colorShifter =
        Tween<double>(begin: 0.0, end: widget.colors.length * _tuning).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(0.0, 1.0, curve: Curves.easeIn),
      ),
    );
    _controller?.forward();
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: AnimatedBuilder(
        animation: _controller,
        builder: (BuildContext context, Widget? child) {
          Shader linearGradient = LinearGradient(colors: widget.colors)
              .createShader(Rect.fromLTWH(0.0, 0.0, _colorShifter.value, 0.0));
          return Opacity(
            opacity: _fadeIn.value != 1.0 ? _fadeIn.value : _fadeOut.value,
            child: Text(
              widget.text,
              style: widget.textStyle != null
                  ? widget.textStyle!.merge(
                      TextStyle(foreground: Paint()..shader = linearGradient))
                  : widget.textStyle,
              textAlign: TextAlign.center,
            ),
          );
        },
      ),
    );
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:splash_screen_view/SplashScreenView.dart';

import 'SecondScreen.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    /// Normal Logo Splash screen
    Widget example1 = SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 200,
      imageSrc: "logo.png",
      backgroundColor: Colors.white,
    );

    /// Logo with animated Colorize text
    Widget example2 = SplashScreenView(
      home: SecondScreen(),
      duration: 5000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Colorize Text",
      textType: TextType.ColorizeAnimationText,
      textStyle: TextStyle(
        fontSize: 40.0,
      ),
      colors: [
        Colors.purple,
        Colors.blue,
        Colors.yellow,
        Colors.red,
      ],
      backgroundColor: Colors.white,
    );

    /// Logo with Typer Animated Text example
    Widget example3 = SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Typer Animated Text",
      textType: TextType.TyperAnimatedText,
      textStyle: TextStyle(
        fontSize: 30.0,
      ),
      backgroundColor: Colors.white,
    );

    /// Logo with Scale Animated Text example
    Widget example4 = SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Scale Animated Text",
      textType: TextType.ScaleAnimatedText,
      textStyle: TextStyle(
        fontSize: 30.0,
      ),
      backgroundColor: Colors.white,
    );

    /// Logo with Normal Text example
    Widget example5 = SplashScreenView(
      home: SecondScreen(),
      duration: 3000,
      imageSize: 100,
      imageSrc: "logo.png",
      text: "Normal Text",
      textStyle: TextStyle(
        fontSize: 30.0,
      ),
      backgroundColor: Colors.white,
    );

    return MaterialApp(
      title: 'Splash screen Demo',
      home: example5,
    );
  }
}

